feature(postgres): Define session role for connection

rldhont opened the pull request:
# Description

Be able to define a session role for postgres connection.

Documentation: https://www.postgresql.org/docs/current/sql-set-role.html

The `session_role` setting will be used to set the current user identifier
of the current SQL session to be `session_role`. Permissions checking for
SQL commands is carried out as though the named role were the one that had
logged in originally.

This is useful to automatically give the ownership of a new object (Table, View, Function) to
the `session_role` group and thus share ownership and associated rights with all members 
of the `session_role` group.

The specified `session_role` must be a role that the current session user
is a member of. (If the session user is a superuser, any role can be selected.)

Funded by 3liz https://3liz.com/

rldhont committed:
feature(postgres): Define session role for connection

Be able to define a session role for postgres connection.

Documentation: https://www.postgresql.org/docs/current/sql-set-role.html

The `session_role` setting will be used to set the current user identifier
of the current SQL session to be `session_role`. Permissions checking for
SQL commands is carried out as though the named role were the one that had
logged in originally.

The specified `session_role` must be a role that the current session user
is a member of. (If the session user is a superuser, any role can be selected.)

nyalldawson commented:
LGTM! Is it possible to test this?

lbartoletti commented on tests/src/providers/testqgspostgresconn.cpp:
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
+      QCOMPARE( result->PQgetvalue( 0, 1 ), "qgis_test_group" );
+      conn->unref();
+
+      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_unkown_group" ) );
```
```suggestion
      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_unknown_group" ) );
```

lbartoletti reviewed (approved):
lgtm

nyalldawson commented on tests/src/providers/testqgspostgresconn.cpp:
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
+
+      const QString sql = QStringLiteral( "SELECT SESSION_USER, CURRENT_USER;" );
+
+      QgsPostgresResult result( conn->PQexec( sql ) );
```
```suggestion
      QgsPostgresResult result = conn->PQexec( sql );
```

nyalldawson commented on tests/src/providers/testqgspostgresconn.cpp:
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
+      uri.removeParam( QStringLiteral( "session_role" ) );
+      conn = QgsPostgresConn::connectDb( uri.connectionInfo( true ), true );
+      QVERIFY( conn != 0 );
+      QgsPostgresResult result( conn->PQexec( sql ) );
```
```suggestion
      QgsPostgresResult result = conn->PQexec( sql );
```

nyalldawson commented on tests/src/providers/testqgspostgresconn.cpp:
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
+      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_group" ) );
+      conn = QgsPostgresConn::connectDb( uri.connectionInfo( true ), true );
+      QVERIFY( conn != 0 );
+      QgsPostgresResult result( conn->PQexec( sql ) );
```
```suggestion
      QgsPostgresResult result = conn->PQexec( sql );
```

nyalldawson commented on tests/src/providers/testqgspostgresconn.cpp:
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
+      QCOMPARE( result.PQgetvalue( 0, 1 ), "qgis_test_user" );
+
+      const char *connstring = getenv( "QGIS_PGTEST_DB" );
+      if ( NULL == connstring ) connstring = "service=qgis_test";
```
Can you fix up all the formatting and old style code here? 
Eg this should be:

```suggestion
      if ( !connstring )
          connstring = "service=qgis_test";
```

nyalldawson commented on tests/src/providers/testqgspostgresconn.cpp:
```diff
@@ -203,6 +204,43 @@ class TestQgsPostgresConn: public QObject
+
+      uri.setParam( QStringLiteral( "session_role" ),  QStringLiteral( "qgis_test_group" ) );
+      conn = QgsPostgresConn::connectDb( uri.connectionInfo( true ), true );
+      QVERIFY( conn != 0 );
```
```suggestion
      QVERIFY( conn );
```

etc

rldhont committed:
Test QgsPostgresConn: fix old style code

nyalldawson commented on src/core/qgsdatasourceuri.h:
```diff
@@ -404,6 +404,8 @@ class CORE_EXPORT QgsDataSourceUri
     //! Generic params store
     QMultiMap<QString, QString> mParams;
+    //! Connection info parameters stored as generic params
+    QStringList mConnectionInfoParams = { "session_role" };
```
This is an ugly approach -- why is it needed here?

rldhont replied:
I need to add `session_role` parameter to `connectionInfo` string. 

I choose this approach to only update 2 methods and to simplify defined new connection information parameters: 
* `connectionInfo` to include `session_role` parameter in the conection information
* `uri` to do not duplicate `session_role` parameter in the uri string

An other approach could be to defined a private class property `mSessionRole` like `mAuthConfigId`, `mUsername` or `mPassword`. With a new private class property,  I have to update 4 methods:
* `setParam`
* `hasParam`
* `param`
* `connectionInfo`

@nyalldawson If you think defined a new private class property is better, I can update my code.

rldhont replied:
I have removed the ugly code.

rldhont commented:
The failing test is not related (SSL network access).

Hi @nyalldawson, do you want me to remove mConnectionInfoParams from QgsDataSourceUri and use a private attribute mSessionRole?

Hi @lbartoletti, I asked you to redo a code review because I had to modify the QgsDataSourceUri class.

rldhont committed:
Tests: Enhance QgsDataSourceUri checkparser with authcfg param

Adding authcfg parameter in the QgsDataSourceUri checkparser test.

Funded by 3liz

rldhont committed:
Tests: Enhance QgsDataSourceUri with setConnection tests

Adding unit tests to the QqsDataSourceUri::setConnection methods.

Funded by 3liz

nyalldawson commented:
I don't think session role handling belongs in QgsDataSourceUri at all -- that's a generic class, and you're putting a postgres specific property in it. That's why there's already the custom parameter API in QgsDataSourceUri. Can't this be used instead of the new member?

rldhont commented:
@nyalldawson ok, I understood, I was not sure how to do it without duplicating too much code. I will remove the changes on QgsDataSourceUri and update the Postgres provider.

nyalldawson commented:
Thanks @rldhont !

rldhont committed:
Tests: Enhance QgsDataSourceUri with connectionInfo tests

nyalldawson commented on src/providers/postgres/qgspostgresconn.h:
```diff
@@ -211,13 +211,35 @@ class QgsPostgresConn : public QObject
+     *
+     * \returns the PostgreSQL connection
      */
     static QgsPostgresConn *connectDb( const QString &connInfo, bool readOnly, bool shared = true, bool transaction = false );
```
Is this still used anywhere? If not, you can safely remove it -- this class isn't stable API

rldhont replied:
It is still used by QgsPostgresTransaction builder derived from QgsTransaction.

nyalldawson commented:
LGTM ! Thanks @rldhont

posikifi commented:
Hi!

I tried to use this "Session ROLE" in QGIS 3.32.2 and I think "Session ROLE"-value is not stored to QGIS profile settings (aka QGIS3.ini).

When I add "Session ROLE"-value in PostGIS connection settings and test connection, it works as expected:

![image](https://github.com/qgis/QGIS/assets/1785962/3ae1c659-b0a6-4346-b060-076dadfdedfc)

When I close "Connection" windows and reopen it, "Session ROLE" value is empty:

![image](https://github.com/qgis/QGIS/assets/1785962/f3e12862-f697-45ff-b3c4-fd6e95016487)

I expect that this "Session ROLE" -value should be stored to QGIS3.ini, but I didn't found anything: 

![image](https://github.com/qgis/QGIS/assets/1785962/37d727b7-4017-47ce-a0a7-ccaf715a77f7)

P
